File: frontend/src/api/feedback.ts
import { api } from './client'

export interface FeedbackAnswer {
  id: string
  nmID: number | null
  userName: string | null
  productValuation: number | null
  text: string | null
  pros: string | null
  cons: string | null
  answer_text: string | null
  is_auto_replied: number | null
  rule_id: number | null
  createdDate: string | null
  updatedDate: string | null
  photoLinks: any[]
  video: Record<string, any> | null
  f_cost: number | null
  bables: string[]
  product_title: string | null
  rule_title: string | null
  [k: string]: unknown
}

export interface FeedbackAnswersParams {
  date_from?: string
  date_to?: string
  nm_id?: number | string
  rating?: number | string
  status?: string
  has_media?: boolean
  paid_only?: boolean
  sort?: string
  order?: string
  page?: number
  page_size?: number
}

export interface FeedbackTag {
  id: number
  tag_text: string
  sentiment: string
  usage_count: number
}

export const feedbackApi = {
  async answers(params: FeedbackAnswersParams): Promise<{
    items: FeedbackAnswer[]
    total: number
    page: number
    page_size: number
    tags_sentiment: Record<string, string>
    rules: Array<{ id: number; title: string }>
  }> {
    const { data } = await api.get('/api/feedback/answers', { params })
    return data
  },
  async tags(filt = 'unclassified'): Promise<FeedbackTag[]> {
    const { data } = await api.get('/api/feedback/tags', { params: { filter: filt } })
    return Array.isArray(data) ? data : []
  },
  async setSentiment(id: number, sentiment: string) {
    const { data } = await api.post(`/api/feedback/tags/${id}/sentiment`, { sentiment })
    return data
  },
  async syncTags(): Promise<{ rows_scanned: number; unique_tags: number; new_tags: number; updated_tags: number }> {
    const { data } = await api.post('/api/feedback/tags/sync')
    return data
  },
}

File: backend/routers/feedback.py
"""Отзывы и ответы — порт WbFeedbackAnswersController (образец: routers/cost.py)."""
from fastapi import APIRouter, Depends, Query
from sqlalchemy.ext.asyncio import AsyncSession

from backend.database import get_db
from backend.deps import get_current_company
from backend.services.feedback_service import FeedbackService

router = APIRouter(prefix="/api/feedback", tags=["feedback"])


@router.get("/answers")
async def feedback_answers(
    date_from: str | None = Query(default=None),
    date_to: str | None = Query(default=None),
    nm_id: int | None = Query(default=None),
    rating: int | None = Query(default=None, ge=1, le=5),
    status: str | None = Query(default=None),
    has_media: bool = Query(default=False),
    paid_only: bool = Query(default=False),
    sort: str = Query(default="createdDate"),
    order: str = Query(default="desc"),
    page: int = Query(default=1, ge=1),
    page_size: int = Query(default=30, ge=1, le=100),
    db: AsyncSession = Depends(get_db),
    company_id: int | None = Depends(get_current_company),
):
    svc = FeedbackService(db, company_id=company_id)
    data = await svc.list(date_from, date_to, nm_id, rating, status,
                          has_media, paid_only, sort, order, page, page_size)
    data["tags_sentiment"] = await svc.tags_sentiment()
    data["rules"] = await svc.rules_options()
    return data


@router.get("/tags")
async def feedback_tags(
    filt: str = Query(default="unclassified", alias="filter"),
    db: AsyncSession = Depends(get_db),
    company_id: int | None = Depends(get_current_company),
):
    """wb_feedback_tags — глобальный справочник, company_id нет (structure.sql:1637)."""
    svc = FeedbackService(db, company_id=company_id)
    return await svc.list_tags("all" if filt == "all" else "unclassified")


@router.post("/tags/{tag_id}/sentiment")
async def feedback_tag_sentiment(
    tag_id: int,
    payload: dict,
    db: AsyncSession = Depends(get_db),
    company_id: int | None = Depends(get_current_company),
):
    svc = FeedbackService(db, company_id=company_id)
    ok = await svc.set_sentiment(tag_id, str(payload.get("sentiment") or ""))
    if not ok:
        from fastapi import HTTPException
        raise HTTPException(status_code=400, detail="Некорректные данные")
    return {"ok": True}


@router.post("/tags/sync")
async def feedback_tags_sync(
    db: AsyncSession = Depends(get_db),
    company_id: int | None = Depends(get_current_company),
):
    svc = FeedbackService(db, company_id=company_id)
    return await svc.sync_tags()
